pkg/scheduler/framework/plugins/volumebinding/volume_binding.go

```diff
@@ -118,27 +121,95 @@ func (pl *VolumeBinding) EventsToRegister() []framework.ClusterEventWithHint {
 		// When CSIStorageCapacity is enabled, pods may become schedulable
 		// on CSI driver & storage capacity changes.
 		{Event: framework.ClusterEvent{Resource: framework.CSIDriver, ActionType: framework.Add | framework.Update}},
-		{Event: framework.ClusterEvent{Resource: framework.CSIStorageCapacity, ActionType: framework.Add | framework.Update}},
+		{Event: framework.ClusterEvent{Resource: framework.CSIStorageCapacity, ActionType: framework.Add | framework.Update}, QueueingHintFn: pl.isSchedulableAfterCSIStorageCapacityChange},
 	}
 	return events
 }
 
+func (pl *VolumeBinding) isSchedulableAfterCSIStorageCapacityChange(logger klog.Logger, pod *v1.Pod, oldObj, newObj interface{}) (framework.QueueingHint, error) {
+	oldCap, newCap, err := util.As[*storagev1.CSIStorageCapacity](oldObj, newObj)
+	if err != nil {
+		return framework.Queue, err
+	}
+
+	logger = klog.LoggerWithValues(
+		logger,
+		"Pod", klog.KObj(pod),
+		"CSIStorageCapacity", klog.KObj(newCap),
+	)
+
+	result, err := processPodVolumesForQHint(pod, func(pvcName string, isEphemeral bool) (bool, error) {
+		pvc, err := pl.PVCLister.PersistentVolumeClaims(pod.Namespace).Get(pvcName)
+		if err != nil {
+			if isEphemeral && apierrors.IsNotFound(err) {
+				err = fmt.Errorf("waiting for ephemeral volume controller to create the persistentvolumeclaim %q", pvcName)
+				return true, err
+			}
+			return true, err
+		}
+
+		className := volume.GetPersistentVolumeClaimClass(pvc)
+		if className == "" {
+			return true, fmt.Errorf("no class for claim %q", pvcName)
+		}
+
+		if newCap.StorageClassName == className {
+			if oldCap == nil {
+				logger.V(4).Info("CSIStorageCapacity was created")
+				return true, nil
+			}
+
+			if newCap.Capacity != oldCap.Capacity {
+				logger.V(4).Info("CSIStorageCapacity referenced by the Pod was created or updated, and changed Capacity", "Capacity", newCap.Capacity)
+				return true, nil
+			}
+
+			if newCap.MaximumVolumeSize != oldCap.MaximumVolumeSize {
+				logger.V(4).Info("CSIStorageCapacity referenced by the Pod was created or updated, and changed MaximumVolumeSize", "MaximumVolumeSize", newCap.MaximumVolumeSize)
+				return true, nil
+			}
+		}
+
+		return false, nil
+	})
+
+	if err != nil {
+		return framework.Queue, err
+	}
+
+	if result {
+		return framework.Queue, nil
+	}
+
+	logger.V(4).Info("CSIStorageCapacity was created or updated, but it doesn't make this pod schedulable")
+	return framework.QueueSkip, nil
+}
+
+func processPodVolumesForQHint(pod *v1.Pod, fn func(string, bool) (bool, error)) (bool, error) {
```

Comment:
IMO, I found the internals of this function difficult to read:

- Since `pod.Spec.Volumes` is no longer needed once the PVC name is obtained, it shouldn't be necessary to process everything within the loop of `pod.Spec.Volumes`.
- The function name `getVolumeClaimName()` also doesn't seem appropriate given the return values.

As done in https://github.com/kubernetes/kubernetes/pull/124996, it seems more readable to obtain the PVC name associated with the Pod and then process it.